PostHog/Surveys/SurveyDisplayController.swift

```diff
@@ -74,20 +81,133 @@
+            response: PostHogSurveyResponse?,
+            responseValues: [String: Any]
+        ) -> NextSurveyQuestion? {
+            guard let response else {
+                hedgeLog("Got response based branching, but missing the actual response.")
+                return nil
+            }
+
+            switch (question, response) {
+            case let (.singleChoice(singleChoiceQuestion), .singleChoice(singleChoiceResponse)):
+                var responseIndex = singleChoiceQuestion.choices.firstIndex(of: singleChoiceResponse ?? "")
+
+                if responseIndex == nil, singleChoiceQuestion.hasOpenChoice == true {
+                    // if the response is not found in the choices, it must be the open choice, which is always the last choice
+                    responseIndex = singleChoiceQuestion.choices.count - 1
+                }
+
+                if let responseIndex, let nextIndex = responseValues["\(responseIndex)"] {
+                    return processBranchingStep(nextIndex: nextIndex, totalQuestions: survey.questions.count)
+                }
+
+                hedgeLog("Could not find response index for specific question.")
+                return nil
+
+            case let (.rating(ratingQuestion), .rating(responseInt)):
+                if let responseInt,
+                   let ratingBucket = getRatingBucketForResponseValue(scale: ratingQuestion.scale, value: responseInt),
+                   let nextIndex = responseValues[ratingBucket]
+                {
+                    return processBranchingStep(nextIndex: nextIndex, totalQuestions: survey.questions.count)
+                }
+                hedgeLog("Could not get response bucket for rating question.")
+                return nil
+
+            default:
+                hedgeLog("Got response based branching for an unsupported question type.")
+                return nil
+            }
         }
+
+        /// Returns next question index based on a branching step result
+        /// - Parameters:
+        ///   - nextIndex: The next index to process
+        ///   - totalQuestions: The total number of questions in the survey
+        /// - Returns: The next question index if found, or nil if not
+        private func processBranchingStep(nextIndex: Any, totalQuestions: Int) -> NextSurveyQuestion? {
+            if let nextIndex = nextIndex as? Int {
+                return .index(min(nextIndex, totalQuestions - 1))
+            }
+            if let nextIndex = nextIndex as? String, nextIndex.lowercased() == "end" {
+                return .end
+            }
+            return nil
+        }
+
+        // Gets the response bucket for a given rating response value, given the scale.
+        // For example, for a scale of 3, the buckets are "negative", "neutral" and "positive".
+        private func getRatingBucketForResponseValue(scale: Int, value: Int) -> String? {
+            switch scale {
+            case 3 where value >= 1 && value <= 3:
```

Comment:
can we move the values to const and name them accordingly
so we do defensive programming for all hard coded values and strings